spec.py

```diff
@@ -403,27 +423,84 @@ def __call__(self, bins=None, bin_width=25.0, normalize=False):
+            self.weights *= fourpi_fraction #!# Additional sampling happens within a restricted domain, which is different depending on the
+                                            #!# energy of the CM. Therefore, an array with values in [0,1] is factored in with the rest of the 
+                                            #!# weights.
 
+        # Compute product energies
+        P1= relscatt.two_body_event(Pa, Pb, self.m1, self.m2, u1)
+        
         # Compute reactivity
-        sigmav = relscatt.get_reactivity(self.ma, self.mb, Pa, Pb, P1, reaction=self.reaction)
+        sigmav = relscatt.get_reactivity(self.ma, self.mb, Pa, Pb, P1, 
+                                         reaction=self.reaction, product_state=self.product_state)   #!#
 
         # Bin all events
-        weights_tot = self.weights * sigmav
+        weights_tot = self.weights * sigmav        
+
+        #!# If a relativistic Doppler shift has to be added on top
+        if self.product_state != 'GS':
+            beta_1 = P1[1:] / P1[0]
+            gamma_1 = (1 - np.sum(beta_1**2,axis=0))**-0.5
+            E_s = self.product_1.E_g  #!# energy at source
+            E_r = E_s / gamma_1 / (1 - np.sum(beta_1*u_prompt_g,axis=0)) #!# energy at receiver
+
+            P1 = relkin.four_vector(E_r, E_r*u_prompt_g)    #!# P1 variable now refers to a different quantity completely! We leave the excited nucleus alone
+                                                            #!# and only care about the gamma from here on.
 
         result = self._make_spec_hist(P1, bins, bin_width, weights_tot, normalize)
 
         return result
 
+    def _sample_u_exNucl(self, Pa, Pb):   #!# Here the magic happens. The sampling in the acceptance cone for the excited nuclei direction is made.
+        p_tot = Pa[1:] + Pb[1:]
+        ortho_e = np.ones(p_tot.shape) * 2**-0.5    #!# A direction which is as little aligned as possible to p_tot has to be picked. This is not 
+                                                    #!# much elegant, but works every time.
+        np.put_along_axis(ortho_e,np.argmax(p_tot,axis=0,keepdims=True),0,axis=0)   #!# See above comment.
+        e1, e2, e3 = tokapart.get_basis_vectors(p_tot, ortho_e)    #!# The tokapart function comes in handy. A minor change has been made to 
+                                                                   #!# make it more versatile and use whatever orthogonal direction to define the 
+                                                                   #!# 3D basis.
+
+        m1_sq,m2_sq = self.m1**2, self.m2**2
+        E_tot_sq = (Pa[0] + Pb[0])**2
+        p_tot_mag_sq = np.linalg.norm(p_tot, axis=0)**2
+        Mtot_sq = E_tot_sq - p_tot_mag_sq
+        Cm_sq = (Mtot_sq + m1_sq - m2_sq)**2
+
+        cos_csi_sq_min = (4*m1_sq*E_tot_sq-Cm_sq) / (4*m1_sq*p_tot_mag_sq)    #!# This comes from kinematics, so energy and momentum conservation.
+        if np.count_nonzero(cos_csi_sq_min>1) > 0:
```

Comment:
What does this line do, exactly? Maybe an explanatory comment would be in order!

Reply:
Let's comment "Should be impossible, but the code is kept in case of exception"

And then 
`Exception("Minimum of cosine of csi is ill-defined.`